### DISCUSSION TITLE
Need help converting fits to jpg

### DISCUSSION THREAD

**Discussion by kopi23** [nom-tam-fits/nom-tam-fits#397] [Q&A]
Hello,

I'm trying to convert fits files (taken with a one shot color camera) to another format like jpg, pngs or tiffs.
I do get results, but they look pretty bad for everything I've tried so far.
Here is what I'm basically doing
```
Fits f = new Fits("image.fits");
ImageHDU hdu = (ImageHDU) f.readHDU();
ImageData imageData = (ImageData) hdu.getData();
short[][] imageFromData = (short[][]) imageData.getData();
```
This gives me a 2d array with the dimensions height x width of the original image with values ranging from -30668 up to 32764.
Then I'm creating a BufferedImage with the dimensions and an ImageType and fill it with the values.
My first attempt was straight forward and this also gives the best result.
```
BufferedImage bImg = new BufferedImage(width, height, BufferedImage.TYPE_BYTE_GRAY);
for (int i = 0; i < imageFromData.length; i++) {
    short[] row = imageFromData[i];
    for (int j = 0; j < row.length; j++) {
    short pixel = row[j];
    bImg.setRGB(j, i, pixel);
}
File ouptut = new File("image.jpg");
ImageIO.write(bimg, "jpg", ouptut);
```
I also tried other ImageTypes than TYPE_BYTE_GRAY and also manipulating the pixel values but so far with no luck.
Best case so far is a gray image with a few brighter small spots where the brightest stars are and the top of the shell.
![IC_443-output](https://user-images.githubusercontent.com/37673792/230746744-15a8703a-ea37-46a8-a514-4e236601aa86.jpg)

I guess I'll have to debayer the image? Then possible stretch it as well?

This is what it looks like when I do it manually.
![IC_443-input](https://user-images.githubusercontent.com/37673792/230746705-4cd4d849-43ff-494f-83ab-640d52965ca0.jpg)

Thanks for any hints, kopi

**Comment by attipaci** [comment]
Hi,

Sorry, I'm not sure I can be helpful. My guess is that `BufferedImage.setRGB()` is not used appropriately in your example. The `setRGB()` calls assumes the 32-bit integer argument is a proper RBGA ColorModel value (such as packing the R byte, G byte, B byte, and alpha channel byte values into a single 4-byte int). However, the FITS `short[][]` image is just a 16-bit integer representation -- most likely of intensity alone. FITS does not have a standard convention to represent color in this way. The typical way to represent RGB colors in FITS would be to store the R, G, and B image planes separately -- either in separate HDUs or as a 3D image cube of [nx][ny][3] dimensions. So, the trick will be knowing how you should map the stored 16-bit values to 3-component RGB or 4-component RGBA values... Good luck!

**Comment by attipaci** [comment]
Without seeing the actual FITS image, I also wonder is there is an issue of signedness at play here. FITS by default supports only signed integer types, but the `BZERO` header keyword can be used to specify a particular 'offset' to signify unsigned data (see the FITS standard document). This Java FITS library does not offer automatic recasting of 'unsigned' `short` values to Java `int`.  So, in your case, I also wonder if the data was meant to range from 0 to 65535, and values &gt;= 32768 are being mapped into negative numbers also. If that is the case, you would have to fix the signedness yourself e.g. via

```java
  ...
  i32[x][y] = ((int) s16[x][y] & 0xffff);
```

**Comment by kopi23** [comment]
Hi,

sorry for the very late reply, have been a few noisy weeks. Really appreciate your time.
I tried to upload the original fits, but the form doesn't support that file type.

Today I finally found time to try a few different things. In particular having a look at the OpenCV.
So currently I'm loading the fits file with nom-tam-fits into an array and then create a Mat (form OpenCV) object from it.
I also tried using the BZERO value from the fits header as an offset but then the image is completely white so I set it manually to the lowest value of the imageFromData array.
```
Header header = f.getHDU(0).getHeader();
int bzero =  header.getIntValue("BZERO");
int offset = 65536 - bzero;
offset = 30668;

Mat src = new Mat(height, width, CvType.CV_8U); //CV_16U
for (int i = 0; i < imageFromData.length; i++) {
  short[] row = imageFromData[i];
  for (int j = 0; j < row.length; j++) {
    src.put(i,j,row[j] + offset);
  }
}
```
From there, I tried debayering it and also equalizing the histogram.
```
Mat rgb = new Mat(src.rows(), src.cols(), src.type());
Imgproc.demosaicing(src, rgb, Imgproc.COLOR_BayerBGGR2RGB);

ArrayList<Mat> splits = new ArrayList<>(3);
Core.split(rgb, splits);

Mat r = new Mat(rgb.rows(), rgb.cols(), rgb.type());
Imgproc.equalizeHist(splits.get(0), r);
Mat g = new Mat(rgb.rows(), rgb.cols(), rgb.type());
Imgproc.equalizeHist(splits.get(1), g);
Mat b = new Mat(rgb.rows(), rgb.cols(), rgb.type());			
Imgproc.equalizeHist(splits.get(2), b);

splits.set(0, r);
splits.set(1, g);
splits.set(2, b);

Mat hst = new Mat(rgb.rows(), rgb.cols(), rgb.type());
Core.merge(splits, hst);

Imgcodecs.imwrite("jelly-hst.png", hst);
```
This is the best result I got till now.
![jelly-hst](https://user-images.githubusercontent.com/37673792/235473477-a07c31a4-8de9-45d9-a80d-4594334a6b6b.png)
All stars are visible and I even got a spec of nebula in color :)

It's far from good but I fear just playing around without knowing what I'm actually doing it won't lead to anything acceptable.
So thanks again for not only reading my post but also taking the time and writing two answers, kopi.

**Reply by rmathar** [reply]
The alternative is to use cfitsio and libtiff (not nom-tam-fits) as implemented in my https://build.opensuse.org/package/show/home:rjmathar/fits2tiff .
The output is TIFF, and I'm using ImageMagick "convert" to move on to other pixel formats (png, jpeg etc)

**Reply by rmathar** [reply]
The code could be improved by setting bzero =0 first (the default) and then taking into account that the ``BZERO`` keyword is not mandatory, so one ought to catch exceptions of the type that ``BZERO`` may be absent. Second, one should read both`` BSCALE`` and ``BZERO`` (and let the scale be 1.0 if absent). Third, the offset added to the raw values should not be 65536-bzero, but bzero itself (and the ``BSCALE`` be used in addition).
Apart from that the discussion can be closed, because questions of converting a grey-scale FITS image into some colored image by furnishing color lookup tables, rescaling the bright or dark pieces &c to let them look "beautiful" are outside the scope of this Java package.

**Comment by attipaci** [comment]
Heeding the suggestion by @rmathar, this discussion shall be closed... :-)
